File: NightNote_25.0708.1.py

import platform
__mainname__ = "NightNote"
__version__ = "25.0708.1"
__author__ = "DONGFANG Lingye"
__email__ = "ly@lingye.online"
def print_help():
    print(f"{__mainname__} version {__version__} by {__author__} <{__email__}>")
    print('''
        Command list:
        p               Print all lines
        i [n]           Insert after line n (end with '.')
        a               Append at end (end with '.')
        d [n]           Delete line n
        w [filename]    Write buffer to file
        o [filename]    Open file and load to buffer
        u l             List buffers
        u c             Clear buffer
        u s [n]         Switch buffer (0~N)
        u r [n]         Remove buffer n
        q               Quit
        h               Show this help
        b               Debug info (OS version, Python version, etc.)
        v               Show about info
        ''')
lines = []  # 用于存储文本的每一行
buffers = [[]]  # 多个缓冲区，初始一个
buffer_files = [""]  # 记录每个缓冲区对应的文件名（如有）
current_buffer = 0  # 当前缓冲区编号

def get_lines():
    return buffers[current_buffer]

def set_lines(new_lines):
    buffers[current_buffer] = new_lines

def set_buffer_file(filename):
    while len(buffer_files) <= current_buffer:
        buffer_files.append("")
    buffer_files[current_buffer] = filename

def get_buffer_file(idx):
    if idx < len(buffer_files):
        return buffer_files[idx]
    return ""
    
def show_buffer_size():
    lines = get_lines()
    total_bytes = sum(len(line.encode('utf-8')) + 1 for line in lines)
    print(f"buffer {current_buffer} {len(lines)} lines {total_bytes} bytes")

def select_file(filename):
    try:
        with open(filename, "r") as f:
            set_lines([line.rstrip('\n') for line in f])
        set_buffer_file(filename)
        print(f"{filename} OK")
    except FileNotFoundError:
        print("E:File not found")
        set_lines([])
        set_buffer_file("")
    except Exception as e:
        print(f"E:Failed to open file: {e}")
        set_lines([])
        set_buffer_file("")
    show_buffer_size()

def p():
    show_buffer_size()
    lines = get_lines()
    for idx, line in enumerate(lines, 1):
        print(f"{idx}: {line}")

def insert(n):
    show_buffer_size()
    lines = get_lines()
    if n < 0 or n > len(lines):
        print("Line number out of range")
        return
    insert_lines = []
    while True:
        text = input()
        if text == ".":
            break
        insert_lines.append(text)
    lines[n:n] = insert_lines
    set_lines(lines)
    show_buffer_size()

def append():
    # 追加内容到末尾
    show_buffer_size()
    lines = get_lines()
    append_lines = []
    while True:
        text = input()
        if text == ".":
            break
        append_lines.append(text)
    lines.extend(append_lines)
    set_lines(lines)
    show_buffer_size()

def delete(n):
    lines = get_lines()
    if n < 1 or n > len(lines):
        print("E:Line number out of range")
        return
    lines.pop(n-1)
    set_lines(lines)
    print(f"Deleted line {n}")
    show_buffer_size()

def write(filename):
    lines = get_lines()
    try:
        with open(filename, "w") as f:
            for line in lines:
                f.write(line + "\n")
        set_buffer_file(filename)
        print(f"OK {filename}")
    except Exception as e:
        print("E:Failed to save:", e)
    show_buffer_size()

def clear_buffer():
    set_lines([])
    print("OK")
    show_buffer_size()

def debug():
    #操作系统版本
    print("DEBUG INFO")
    print(__version__)
    print("OS:", platform.system())
    print("OS Version:", platform.version())
    print("Platform:", platform.platform())
    #py版本
    print("Python Version:", platform.python_version())
    show_buffer_size()

def about():
    print(f"{__mainname__} version {__version__} by {__author__} <{__email__}>")
    print("A simple line editor, similar to Linux's ed editor.")
    print("Type 'h' for help.")
    print("Type 'q' to quit.")

def switch_buffer(n):
    global current_buffer
    if n < 0:
        print("E:Buffer number must be >= 0")
        return
    while n >= len(buffers):
        buffers.append([])
    while n >= len(buffer_files):
        buffer_files.append("")
    current_buffer = n
    show_buffer_size()

def list_buffers():
    print("Buffers info:")
    for idx, buf in enumerate(buffers):
        fname = get_buffer_file(idx)
        total_bytes = sum(len(line.encode('utf-8')) + 1 for line in buf)
        mark = "    <--" if idx == current_buffer else ""
        print(f"[{idx}] {len(buf)} lines, {total_bytes} bytes, '{fname}' {mark}")

def remove_buffer(n):
    global current_buffer
    if n < 0 or n >= len(buffers):
        print("E:Buffer number out of range")
        return
    if len(buffers) == 1:
        print("E:At least one buffer must remain")
        return
    buffers.pop(n)
    buffer_files.pop(n)
    if current_buffer == n:
        current_buffer = 0
        print("OK:switch to buffer 0")
    elif current_buffer > n:
        current_buffer -= 1
    print(f"Buffer {n} removed.")

def main():
    print(f"{__mainname__} {__version__}")
    while True:
        cmd = input("]").strip()
        if cmd == "p":
            p()
        elif cmd.startswith("i "):
            try:
                n = int(cmd.split()[1])
                insert(n)
            except:
                print("i n  Insert after line n (end with '.')")
                
        elif cmd == "a":
            append()
        elif cmd.startswith("d "):
            try:
                n = int(cmd.split()[1])
                delete(n)
            except:
                print("d n  Delete line n")
                
        elif cmd.startswith("w "):
            filename = cmd[2:].strip()
            write(filename)
            
        elif cmd.startswith("o "):
            filename = cmd[2:].strip()
            if filename:
                select_file(filename)
            else:
                print("o filename   Open file and load to buffer")
                
        elif cmd == "c":
            clear_buffer()
        
        elif cmd.startswith("s "):
            try:
                n = int(cmd.split()[1])
                switch_buffer(n)
            except:
                print("s n  Switch buffer (0~N)")
        elif cmd.startswith("r "):
            try:
                n = int(cmd.split()[1])
                remove_buffer(n)
            except:
                print("r n  Remove buffer n")
                
        elif cmd.startswith("u "):
            args = cmd.split()
            if len(args) == 2 and args[1] == "c":
                clear_buffer()
            elif len(args) == 2 and args[1] == "l":
                list_buffers()
            elif len(args) == 3 and args[1] == "s":
                try:
                    n = int(args[2])
                    switch_buffer(n)
                except:
                    print("u s [n]  Switch buffer (0~N)")
            elif len(args) == 3 and args[1] == "r":
                try:
                    n = int(args[2])
                    remove_buffer(n)
                except:
                    print("u r n  Remove buffer n")
            else:
                print("u c  Clear buffer | u s [n]  Switch buffer | u r [n]  Remove buffer n")
                
        elif cmd == "l":
            list_buffers()
        elif cmd == "q":
            print("Bye!")
            break
        elif cmd == "h":
            print_help()
        elif cmd == "b":
            debug()
        elif cmd == "v":
            about()
        elif cmd == "c":
            clear_buffer()
        else:
            print("E:Unknown command, enter h for help.")
            show_buffer_size()

if __name__ == "__main__":
    main()

File: NightNote_25.0708.1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINES 10000
#define MAX_BUFFERS 8
#define MAX_LINE_LEN 1024
#define MAX_FILENAME 256

const char* MAINNAME = "NightNote";
const char* VERSION = "25.0708.1";
const char* AUTHOR = "DONGFANG Lingye";
const char* EMAIL = "ly@lingye.online";

char buffers[MAX_BUFFERS][MAX_LINES][MAX_LINE_LEN];
int buffer_line_counts[MAX_BUFFERS] = {0};
char buffer_files[MAX_BUFFERS][MAX_FILENAME] = {""};
int current_buffer = 0;

void print_help() {
    printf("%s version %s by %s <%s>\n", MAINNAME, VERSION, AUTHOR, EMAIL);
    printf(
        "Command list:\n"
        "p               Print all lines\n"
        "i [n]           Insert after line n (end with '.')\n"
        "a               Append at end (end with '.')\n"
        "d [n]           Delete line n\n"
        "w [filename]    Write buffer to file\n"
        "o [filename]    Open file and load to buffer\n"
        "u l             List buffers\n"
        "u c             Clear buffer\n"
        "u s [n]         Switch buffer (0~N)\n"
        "u r [n]         Remove buffer n\n"
        "q               Quit\n"
        "h               Show this help\n"
        "b               Debug info\n"
        "v               Show about info\n"
    );
}

void show_buffer_size() {
    int lines = buffer_line_counts[current_buffer];
    int total_bytes = 0;
    for (int i = 0; i < lines; ++i) {
        total_bytes += strlen(buffers[current_buffer][i]) + 1;
    }
    printf("buffer %d %d lines %d bytes\n", current_buffer, lines, total_bytes);
}

void p() {
    show_buffer_size();
    int lines = buffer_line_counts[current_buffer];
    for (int i = 0; i < lines; ++i) {
        printf("%d: %s\n", i + 1, buffers[current_buffer][i]);
    }
}

void insert(int n) {
    show_buffer_size();
    int lines = buffer_line_counts[current_buffer];
    if (n < 0 || n > lines) {
        printf("Line number out of range\n");
        return;
    }
    if (lines >= MAX_LINES) {
        printf("Buffer is full, cannot insert more lines.\n");
        return;
    }
    char temp[MAX_LINE_LEN];
    int insert_count = 0;
    int max_insert = MAX_LINES - lines; // 剩余可插入行数
    char insert_lines[MAX_LINE_LEN][MAX_LINE_LEN]; // 每次插入不可能超过MAX_LINE_LEN行
    while (insert_count < max_insert) {
        if (!fgets(temp, sizeof(temp), stdin)) break;
        temp[strcspn(temp, "\n")] = 0;
        if (strcmp(temp, ".") == 0) break;
        strncpy(insert_lines[insert_count++], temp, MAX_LINE_LEN - 1);
        insert_lines[insert_count-1][MAX_LINE_LEN-1] = '\0';
    }
    if (insert_count == 0) return;
    // 移动原有行
    for (int i = lines - 1; i >= n; --i) {
        strncpy(buffers[current_buffer][i + insert_count], buffers[current_buffer][i], MAX_LINE_LEN - 1);
        buffers[current_buffer][i + insert_count][MAX_LINE_LEN-1] = '\0';
    }
    // 插入新行
    for (int i = 0; i < insert_count; ++i) {
        strncpy(buffers[current_buffer][n + i], insert_lines[i], MAX_LINE_LEN - 1);
        buffers[current_buffer][n + i][MAX_LINE_LEN-1] = '\0';
    }
    buffer_line_counts[current_buffer] += insert_count;
    show_buffer_size();
}

void append() {
    show_buffer_size();
    int lines = buffer_line_counts[current_buffer];
    char temp[MAX_LINE_LEN];
    while (1) {
        if (!fgets(temp, sizeof(temp), stdin)) break;
        temp[strcspn(temp, "\n")] = 0;
        if (strcmp(temp, ".") == 0) break;
        strncpy(buffers[current_buffer][lines++], temp, MAX_LINE_LEN);
    }
    buffer_line_counts[current_buffer] = lines;
    show_buffer_size();
}

void delete(int n) {
    int lines = buffer_line_counts[current_buffer];
    if (n < 1 || n > lines) {
        printf("E:Line number out of range\n");
        return;
    }
    for (int i = n - 1; i < lines - 1; ++i) {
        strncpy(buffers[current_buffer][i], buffers[current_buffer][i + 1], MAX_LINE_LEN);
    }
    buffer_line_counts[current_buffer]--;
    printf("Deleted line %d\n", n);
    show_buffer_size();
}

void write(const char* filename) {
    FILE* f = fopen(filename, "w");
    if (!f) {
        printf("E:Failed to save: %s\n", filename);
        show_buffer_size();
        return;
    }
    int lines = buffer_line_counts[current_buffer];
    for (int i = 0; i < lines; ++i) {
        fprintf(f, "%s\n", buffers[current_buffer][i]);
    }
    fclose(f);
    strncpy(buffer_files[current_buffer], filename, MAX_FILENAME);
    printf("OK %s\n", filename);
    show_buffer_size();
}

void select_file(const char* filename) {
    FILE* f = fopen(filename, "r");
    if (!f) {
        printf("E:File not found\n");
        buffer_line_counts[current_buffer] = 0;
        buffer_files[current_buffer][0] = '\0';
        show_buffer_size();
        return;
    }
    int count = 0;
    char temp[MAX_LINE_LEN];
    while (fgets(temp, sizeof(temp), f) && count < MAX_LINES) {
        temp[strcspn(temp, "\n")] = 0;
        strncpy(buffers[current_buffer][count++], temp, MAX_LINE_LEN);
    }
    fclose(f);
    buffer_line_counts[current_buffer] = count;
    strncpy(buffer_files[current_buffer], filename, MAX_FILENAME);
    printf("%s OK\n", filename);
    show_buffer_size();
}

void clear_buffer() {
    buffer_line_counts[current_buffer] = 0;
    printf("OK\n");
    show_buffer_size();
}

void debug() {
    printf("DEBUG INFO\n");
    printf("%s\n", VERSION);
    printf("OS: Linux (C version, no platform details)\n");
    show_buffer_size();
}

void about() {
    printf("%s version %s by %s <%s>\n", MAINNAME, VERSION, AUTHOR, EMAIL);
    printf("A simple line editor, similar to Linux's ed editor.\n");
    printf("Type 'h' for help.\n");
    printf("Type 'q' to quit.\n");
}

void switch_buffer(int n) {
    if (n < 0 || n >= MAX_BUFFERS) {
        printf("E:Buffer number must be >= 0 and < %d\n", MAX_BUFFERS);
        return;
    }
    current_buffer = n;
    show_buffer_size();
}

void list_buffers() {
    printf("Buffers info:\n");
    for (int i = 0; i < MAX_BUFFERS; ++i) {
        int lines = buffer_line_counts[i];
        int total_bytes = 0;
        for (int j = 0; j < lines; ++j) {
            total_bytes += strlen(buffers[i][j]) + 1;
        }
        printf("[%d] %d lines, %d bytes, '%s'%s\n", i, lines, total_bytes, buffer_files[i], (i == current_buffer) ? "    <--" : "");
    }
}

void remove_buffer(int n) {
    if (n < 0 || n >= MAX_BUFFERS) {
        printf("E:Buffer number out of range\n");
        return;
    }
    if (MAX_BUFFERS == 1) {
        printf("E:At least one buffer must remain\n");
        return;
    }
    // 这里只是清空，不真正移除
    buffer_line_counts[n] = 0;
    buffer_files[n][0] = '\0';
    if (current_buffer == n) {
        current_buffer = 0;
        printf("OK:switch to buffer 0\n");
    }
    printf("Buffer %d removed.\n", n);
}

int main() {
    printf("%s %s\n", MAINNAME, VERSION);
    char cmd[256];
    while (1) {
        printf("]");
        if (!fgets(cmd, sizeof(cmd), stdin)) break;
        cmd[strcspn(cmd, "\n")] = 0;
        if (strcmp(cmd, "p") == 0) {
            p();
        } else if (strncmp(cmd, "i ", 2) == 0) {
            int n = atoi(cmd + 2);
            insert(n);
        } else if (strcmp(cmd, "a") == 0) {
            append();
        } else if (strncmp(cmd, "d ", 2) == 0) {
            int n = atoi(cmd + 2);
            delete(n);
        } else if (strncmp(cmd, "w ", 2) == 0) {
            write(cmd + 2);
        } else if (strncmp(cmd, "o ", 2) == 0) {
            select_file(cmd + 2);
        } else if (strcmp(cmd, "c") == 0) {
            clear_buffer();
        } else if (strncmp(cmd, "s ", 2) == 0) {
            int n = atoi(cmd + 2);
            switch_buffer(n);
        } else if (strncmp(cmd, "r ", 2) == 0) {
            int n = atoi(cmd + 2);
            remove_buffer(n);
        } else if (strncmp(cmd, "u ", 2) == 0) {
            if (strcmp(cmd + 2, "c") == 0) {
                clear_buffer();
            } else if (strcmp(cmd + 2, "l") == 0) {
                list_buffers();
            } else if (strncmp(cmd + 2, "s ", 2) == 0) {
                int n = atoi(cmd + 4);
                switch_buffer(n);
            } else if (strncmp(cmd + 2, "r ", 2) == 0) {
                int n = atoi(cmd + 4);
                remove_buffer(n);
            } else {
                printf("u c  Clear buffer | u s [n]  Switch buffer | u r [n]  Remove buffer n\n");
            }
        } else if (strcmp(cmd, "l") == 0) {
            list_buffers();
        } else if (strcmp(cmd, "q") == 0) {
            printf("Bye!\n");
            break;
        } else if (strcmp(cmd, "h") == 0) {
            print_help();
        } else if (strcmp(cmd, "b") == 0) {
            debug();
        } else if (strcmp(cmd, "v") == 0) {
            about();
        } else {
            printf("E:Unknown command, enter h for help.\n");
            show_buffer_size();
        }
    }
    return 0;
}